myhpom/tests/test_irods_view.py

```diff
@@ -0,0 +1,84 @@
+        proc_mock = MagicMock()
+        type(proc_mock).stdout = PropertyMock(return_value='content')
+        return proc_mock
+
+
+@patch('myhpom.views.irods.GLOBAL_SESSION')
+class IrodsDownloadTest(TestCase):
+    url = reverse('myhpom:irods_download', kwargs={'path': 'a_path'})
+
+    def test_non_existant(self, session_mock):
+        # If the path does not exist in irods, return an error
+        session_mock.run.side_effect = SessionException(1, '', '')
+        response = self.client.get(self.url)
+        self.assertEqual(404, response.status_code)
+
+    def test_file(self, session_mock):
+        # When the file is in the system, but isn't an AdvanceDirective,
+        # any user should be able to download the file (regardless of whether
+        # they are logged in or not)
+        session_mock.run.side_effect = run_side_effect
+        session_mock.run_safe.side_effect = run_safe_side_effect
+        response = self.client.get(self.url)
+        self.assertEqual(''.join(response.streaming_content), 'content')
+
+    def test_mime_type(self, session_mock):
+        # The content type is populated if possible - a rando filename will
+        # result in octet-stream:
+        session_mock.run.side_effect = run_side_effect
+        session_mock.run_safe.side_effect = run_safe_side_effect
+        response = self.client.get(self.url)
+        self.assertEqual('application-x/octet-stream', response.get('content-type'))
+
+        # A PDF file will have the correct content type:
+        self.url = reverse('myhpom:irods_download', kwargs={'path': 'a_path.pdf'})
+        response = self.client.get(self.url)
+        self.assertEqual('application/pdf', response.get('content-type'))
+
+    def test_advance_directive(self, session_mock):
+        # When the file is an AdvanceDirective, the user must be logged in and
+        # they must own the file in order to download it.
+        session_mock.run.side_effect = run_side_effect
+        session_mock.run_safe.side_effect = run_safe_side_effect
+        user = UserFactory()
+        user.set_password('password')
+        user.save()
+
+        # If the user is not logged in, then return an error
+        advancedirective = AdvanceDirective(user=user, valid_date=now(), document='a_path', share_with_ehs=False)
+        advancedirective.save()
+        response = self.client.get(self.url)
+        self.assertEqual(404, response.status_code)
+
+        # When the user is logged in, they can retrieve their file:
+        self.assertTrue(self.client.login(username=user.username, password='password'))
+        response = self.client.get(self.url)
+        self.assertEqual(''.join(response.streaming_content), 'content')
+
+        # When the user doesn't own the path, return an error
+        another_user = UserFactory()
+        advancedirective.user = another_user
+        advancedirective.save()
+        response = self.client.get(self.url)
+        self.assertEqual(404, response.status_code)
```

Comment:
This test passes on my system, BUT: If the user deletes their advance directive, it leaves it on the filesystem. So the above comment is still valid: 

1. User 1 uploads MYAD.pdf, which is saved with that filename. 
2. User 2 uploads MYAD.pdf, which is saved as MYAD_RANDOM.pdf.
3. User 1 "deletes" their AD, but MYAD.pdf remains in the irods filesystem. 
4. Now User 2, who sees MYAD_RANDOM.pdf, can download MYAD.pdf.

Reply:
Good catch!

From what I can tell its not the responsibility of the Storage drivers to go back and delete a file when a model is deleted (https://docs.djangoproject.com/en/1.11/ref/models/fields/#django.db.models.fields.files.FieldFile.delete -- which makes some sense...). I've heard of two ways around this in django:

 * make a cron job that, for each model with a FileField you care about, computes the set of files that should exist...and then delete any that aren't in that set. Slow, bet reliable.
 * use pre_save/post_delete hooks on the models to delete the file. Fast, but could fail for bulk operations (if we used them).

I suggest...I'm not sure? The second option (add pre_save/post_delete hooks to AdvanceDirective)?

Reply:
I think I would do #2 and then, if we ever implement bulk ops, swap it to the more complicated #1.

Reply:
Okay, I'll start with that then.